ond.py

```diff
@@ -86,30 +86,28 @@ def run_protocol(self, config_):
                 )
 
                 if file_list is not None:
-                    features_dict, logits_dict = algorithm.execute(
-                        "FeatureExtraction", file_list
+                    features_dict, logits_dict = algorithm.feature_extraction(
+                        file_list
```

Comment:
Maybe I am missing something but shouldn't the parameter here be toolset?

Reply:
With this port, I'm aiming to get away from the use of `toolset`, so for the time being am passing parameters directly. I'm guessing some of the mechanics of this will change as things evolve but am trying to keep things simple for now.

Reply:
The main issue would that the port wouldn't support the different signatures that algorithms have. The reason for using toolset as opposed to `file_list` and `dataset_root` is because algorithms can have different signatures. A simple example would be that the baseline does not need `feature_dict` in `novelty_classification`. However, some algorithms need features because they are using an additional module on the top of the network to reason about sample-wise novelty. Thus by using the toolset as a parameter to the calls we defer the input/output requirements to the algorithm developer rather than trying to address them in the protocol. We had also started with exact calls to the algorithms methods in the client and ran into this problem during the evaluation where algorithms have slightly different requirements and they have to addressed somewhere. Essentially adapter provides that interface where the different requirements are addressed. Additionally, as it is packaged with the algorithms, it is easy to read through those requirements. Rather than going through the protocol that is trying to address them but is not present in the same package.

Reply:
I always thought that the algorithm adapters would solve exactly this problem. Is that not the case? Right now, all algorithm developers have to follow this API of stringly-typed algorithm phases. Instead could we just develop algorithm adapters that pass the required values along? If not, we can stick with this stringly-typed mechanism, but it has a lot of disadvantages and we should work hard to avoid it in favor of something else.

Reply:
Based on my understanding the stringly-types mechanism is different from the toolset. String typed mechanism was `execute` with `step_descriptor` which was used to determine which function was executing. I agree that we shouldn't be using the `execute` method. The adapters are doing just that, passing required values to the algorithms. Since different algorithms can have different required values, a toolset is passed to the adapter so that the adapter can fulfill the requirements of the algorithm.

Reply:
> @as6520 @waxlamp Please see latest commits. At the risk of introducing too many changes in one PR, I went ahead and updated the dataset_request logic to closely follow that of the original OND protocol because it affects the inputs to the algorithm step calls that this PR originally sought to address.
> 
> With these changes, I've maintained the idea of the `toolset` but propose replacing `toolset` with three dicts that more clearly define what's being supplied/returned/used by the algorithm or adapter at each step:
> 
> * `algo_config_params`: Configuration parameters needed to initialize the algorithm/detector; in other words, parameters that shouldn't change between rounds or probably even between tests.
> * `algo_test_params`: Test-related parameters and information, e.g., dataset ids, dataset root, filename of red light image (if applicable).
> * `algo_test_data`: Algorithm input and/or output for the various steps during a test, e.g., from feature extraction, novelty classification, etc.
> 
> This can be refined in upcoming PRs, but looking for feedback on whether this is an appropriate step in that direction.

This does indeed look like a good approach to me. The only thing I would add is type hints to the various "toolset" values--with those, this code will become a lot easier to digest for someone else reading it. (The algorithm adapters being used should also be type hinted to provide insight into that part of the code.)

If you're not sure how to go about this, we can work on it this week during our mini hackathon.